pkg/spanfilter/splitpolicy.go

```diff
@@ -0,0 +1,115 @@
+package spanfilter
+
+import (
+	"fmt"
+
+	"github.com/grafana/tempo/pkg/spanfilter/config"
+	"github.com/grafana/tempo/pkg/spanfilter/policymatch"
+	v1 "github.com/grafana/tempo/pkg/tempopb/resource/v1"
+	tracev1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/traceql"
+	"golang.org/x/exp/maps"
+)
+
+// splitPolicy is the result of parsing a policy from the config file to be
+// specific about the area the given policy is applied to.
+type splitPolicy struct {
+	// ResourceMatch is a set of resource attributes that must match a span for the span to match the policy.
+	ResourceMatch *policymatch.AttributePolicyMatch
+	// SpanMatch is a set of span attributes that must match a span for the span to match the policy.
+	SpanMatch *policymatch.AttributePolicyMatch
+	// IntrinsicMatch is a set of intrinsic attributes that must match a span for the span to match the policy.
+	IntrinsicMatch *policymatch.IntrinsicPolicyMatch
+}
+
+func newSplitPolicy(policy *config.PolicyMatch) (*splitPolicy, error) {
+	var resourceAttributeFilters []policymatch.AttributeFilter
+	var spanAttributeFilters []policymatch.AttributeFilter
+	var intrinsicFilters []policymatch.IntrinsicFilter
+
+	for _, pa := range policy.Attributes {
+		attr, err := traceql.ParseIdentifier(pa.Key)
+		if err != nil {
+			return nil, err
+		}
+
+		if attr.Intrinsic > 0 {
+			var filter policymatch.IntrinsicFilter
+			if policy.MatchType == config.Strict {
+				switch attr.Intrinsic {
+				case traceql.IntrinsicKind:
+					switch v := pa.Value.(type) {
+					case tracev1.Span_SpanKind:
+						filter = policymatch.NewKindIntrinsicFilter(v)
+					case string:
+						if kind, ok := tracev1.Span_SpanKind_value[v]; ok {
+							filter = policymatch.NewKindIntrinsicFilter(tracev1.Span_SpanKind(kind))
+						} else {
+							return nil, fmt.Errorf("currently unsupported kind intrinsic string value: %s; supported values: %v", v, maps.Keys(tracev1.Span_SpanKind_value))
+						}
+					default:
+						return nil, fmt.Errorf("invalid kind intrinsic value: %v", v)
+					}
+				case traceql.IntrinsicStatus:
+					switch v := pa.Value.(type) {
+					case tracev1.Status_StatusCode:
+						filter = policymatch.NewStatusIntrinsicFilter(v)
+					case string:
+						if code, ok := tracev1.Status_StatusCode_value[v]; ok {
+							filter = policymatch.NewStatusIntrinsicFilter(tracev1.Status_StatusCode(code))
+						} else {
+							return nil, fmt.Errorf("currently unsupported status intrinsic string value: %s; supported values: %v", v, maps.Keys(tracev1.Status_StatusCode_value))
+						}
+					default:
+						return nil, fmt.Errorf("currently unsupported intrinsic: %v", v)
+					}
+				case traceql.IntrinsicName:
+					filter = policymatch.NewNameIntrinsicFilter(pa.Value.(string))
+				}
+			} else {
+				filter, err = policymatch.NewRegexpIntrinsicFilter(attr.Intrinsic, pa.Value.(string))
```

Comment:
we should avoid the unsafe cast `pa.Value.(string)`. I believe it would be possible to type an integer into the config and panic the application. let's do the ol' safe cast: `s, ok := pa.Value.(string)`